No associated issue : add arrow-up button for Link bottom to top page

Nutchawat opened the pull request:
No associated issue : I've add arrow-up button when click it will scroll to top (Link to section herounit)
![screen shot 2560-05-22 at 5 42 16 pm](https://cloud.githubusercontent.com/assets/4596685/26304560/12436a20-3f16-11e7-905d-a79342bf862d.png)

n3tr commented on package.json:
```diff
@@ -37,6 +37,7 @@
   },
   "dependencies": {
     "express": "^4.15.3",
+    "jquery": "^3.2.1",
```
Do we need to include `jquery` here?

dtinth replied:
Please remove jQuery ;)

Nutchawat commented:
no i forget to remove it. sorry

On Mon, May 22, 2017 at 6:11 PM, Jirat Ki. <notifications@github.com> wrote:

> *@n3tr* commented on this pull request.
> ------------------------------
>
> In package.json
> <https://github.com/reactbkk/2.0.0/pull/119#discussion_r117719520>:
>
> > @@ -37,6 +37,7 @@
>    },
>    "dependencies": {
>      "express": "^4.15.3",
> +    "jquery": "^3.2.1",
>
> Do we need to include jquery here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/reactbkk/2.0.0/pull/119#pullrequestreview-39447702>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEYjzXFXgjsom_20oL9nMMgkBNcSvelyks5r8W14gaJpZM4NiLyj>
> .
>

n3tr commented:
Arrow image looks blurry on my mac (Retina Display), Can we replace it with svg?
![](https://cloud.githubusercontent.com/assets/155392/26305717/b85e9282-3f1a-11e7-83a0-0755f5d5210d.png)

Also, I don't think we need to show `up-to-top` arrow if we on top of the page.

Nutchawat commented:
I don't have any svg but I could find it.

the hidden event up-to-top when on top page is doing krab.

On Mon, May 22, 2017 at 6:17 PM, Jirat Ki. <notifications@github.com> wrote:

> Arrow image looks blurry on my mac (Retina Display), Can we replace it
> with svg?
>
> <https://cloud.githubusercontent.com/assets/155392/26305717/b85e9282-3f1a-11e7-83a0-0755f5d5210d.png>
>
> Also, I don't think we need to show up-to-top arrow if we on top of the
> page.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/reactbkk/2.0.0/pull/119#issuecomment-303071597>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEYjzVJUXMrZTmAByKjDuvB-F4gdLobzks5r8W6_gaJpZM4NiLyj>
> .
>

Nutchawat commented:
I've push arrow-up.svg and try to track documents scroll with react-scroll
events.

I think I have to register event on componentDidMount but I'm not sure how to do that.
Please advise me.

On Mon, May 22, 2017 at 6:56 PM, Nutchawat Ploywattanawong <
pnutchaw@gmail.com> wrote:

> I don't have any svg but I could find it.
>
> the hidden event up-to-top when on top page is doing krab.
>
> On Mon, May 22, 2017 at 6:17 PM, Jirat Ki. <notifications@github.com>
> wrote:
>
>> Arrow image looks blurry on my mac (Retina Display), Can we replace it
>> with svg?
>>
>> <https://cloud.githubusercontent.com/assets/155392/26305717/b85e9282-3f1a-11e7-83a0-0755f5d5210d.png>
>>
>> Also, I don't think we need to show up-to-top arrow if we on top of the
>> page.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/reactbkk/2.0.0/pull/119#issuecomment-303071597>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AEYjzVJUXMrZTmAByKjDuvB-F4gdLobzks5r8W6_gaJpZM4NiLyj>
>> .
>>
>
>

dtinth commented:
@Nutchawat Please check the implementation of a mobile menu in #112 for an example of how this can be implemented. In particular, `MobileNavigation.js` lines 21~53.

Nutchawat commented:
Binding event scroll for handle arrow-up hidden when scroll to top.
@dtinth 
Following componentDidMount's implemented from MobileNavigation.js as your advised.

Please check the issue.

Thank you krab

Nutchawat commented:
Binding event scroll for handle arrow-up hidden when scroll to top has
finished.
@dtinth <https://github.com/dtinth>
Following componentDidMount's implemented from MobileNavigation.js as your
advised.

Please check the issue.

Thank you krab

On Tue, May 23, 2017 at 12:33 PM, reactbkkbot <notifications@github.com>
wrote:

> @Nutchawat <https://github.com/nutchawat> I found an issue tag now, thank
> you.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/reactbkk/2.0.0/pull/119#issuecomment-303294798>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEYjzUVw_SxenKaJFf31GLDfuiKjrFCKks5r8m-fgaJpZM4NiLyj>
> .
>

Nutchawat commented:
@dtinth <https://github.com/dtinth>
<https://github.com/dtinth>ตอนผม forked มาแก้ไข run แล้วไม่ error อะไร พอ
merge ลง master แล้ว conflict ไฟล์ index.js
ไม่ทราบว่าควรแก้ไขอย่างไรครับ

ขอบคุณครับ

On Tue, May 23, 2017 at 2:04 PM, Nutchawat Ploywattanawong <
pnutchaw@gmail.com> wrote:

> Binding event scroll for handle arrow-up hidden when scroll to top has
> finished.
> @dtinth <https://github.com/dtinth>
> Following componentDidMount's implemented from MobileNavigation.js as your
> advised.
>
> Please check the issue.
>
> Thank you krab
>
> On Tue, May 23, 2017 at 12:33 PM, reactbkkbot <notifications@github.com>
> wrote:
>
>> @Nutchawat <https://github.com/nutchawat> I found an issue tag now,
>> thank you.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/reactbkk/2.0.0/pull/119#issuecomment-303294798>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AEYjzUVw_SxenKaJFf31GLDfuiKjrFCKks5r8m-fgaJpZM4NiLyj>
>> .
>>
>
>

dtinth commented on components/BacktoTopSection.js:
```diff
@@ -0,0 +1,62 @@
+  constructor (props) {
+    super(props)
+    this.state = {
+      cssHide: 'button-hide'
```
`cssHide` is actually a derived state. Right now the code looks like this:

```
           change        bind
handleScroll ==> cssHide <== top arrow
```

This couples the code that handles the scrolling to the CSS class. This makes it harder when we want to add more complex logic to our view (e.g. if we also want to hide the button on desktop devices, for example). This can lead to inconsistent UI state.

But in React, we try to make our code reactive by storing ‘facts,’ and compute the desired state of the view on `render`.

Let’s try to define the meaning by asking few questions…

1. `cssHide` is the CSS class to apply to the top-arrow.

   What CSS class should we apply?

2. We should apply `'button-hide'` class when button should be hidden. Otherwise, we should not apply this class.

   When should the button be hidden?

3. The button should be hidden when the user is at the top of the page. This fact is controlled by the user.

So in this case, I think this is the data that we should store in React state.

```js
  atTopOfPage: false
```

Then on `render`, we can then do this:

```js
  return <div
    id='top-arrow'
    className={`top-arrow ${this.state.atTopOfPage ? 'button-hide' : ''}`}
    onClick={this.scrollToTop}
  />
```

Now we know what class names could be applied to `top-arrow` by looking at the `div` tag only.

dtinth commented on components/BacktoTopSection.js:
```diff
@@ -0,0 +1,62 @@
+
+  handleScroll = () => {
+    if (document.body.scrollTop > 20 || document.documentElement.scrollTop > 20) {
+      this.setState({ cssHide: '' })
```
Because this function is called on every scrolling, we should call `setState` only when the state really needs to change.

```js
    if (this.state.atTopOfPage) this.setState({ atTopOfPage: false })
```

dtinth commented on components/BacktoTopSection.js:
```diff
@@ -0,0 +1,62 @@
+        <style jsx>{`
+          .top-arrow {
+            position: fixed;
+            left: auto;
```
This is default, not needed.

dtinth commented on components/BacktoTopSection.js:
```diff
@@ -0,0 +1,62 @@
+            background: #222 url(static/arrow-up.svg) no-repeat center center;
+            background-size: 30px 20px;
+            border-radius: 60px;
+            -webkit-transition: 1s;
```
styled-jsx already adds prefixes for you, so `-webkit-transition` and `-moz-transition` is not needed.

dtinth commented on components/BacktoTopSection.js:
```diff
@@ -0,0 +1,62 @@
+            background-color: #777;
+          }
+          .button-hide {
+            display: none;
```
Instead of using `display: none;` which causes button to immediately hide, let’s change the `opacity` so that it can be animated.

Also, let’s add `pointer-events: none;` to make sure that user cannot click on it.

dtinth commented:
Please fix the merge conflicts, thank you.

dtinth commented on pages/index.js:
```diff
@@ -5,6 +5,7 @@ import SectionSeparator from '../components/SectionSeparator'
 import SpeakersSection from '../components/SpeakersSection'
 import TicketsSection from '../components/TicketsSection'
 import SponsorsSection from '../components/SponsorsSection'
+import BacktoTopSection from '../components/BacktoTopSection'
```
This is not a section; it is a button. Let's call it BackToTopButton.

dtinth commented on components/BacktoTopSection.js:
```diff
@@ -0,0 +1,60 @@
+  }
+
+  handleScroll = () => {
+    if (document.body.scrollTop > 20 || document.documentElement.scrollTop > 20) {
```
We can use the `window.scrollY` instead. It is supported in most modern browsers.

dtinth commented on components/BacktoTopSection.js:
```diff
@@ -0,0 +1,60 @@
+  }
+
+  componentDidMount () {
+    window.addEventListener('scroll', this.handleScroll)
```
This handler will not be called when the page is first loaded. So we also need to call it manually on `componentDidMount`:

```js
    this.handleScroll()
```

Nutchawat commented:
@dtinth 
Sry for slowly interactive, I've already fixed all of your change request krab.
1. change ``state.cssHide`` to ``state.atTopOfPage``
2. setState only one time call when state.atTopOfPage really need change: 
``if (this.state.atTopOfPage) this.setState({ atTopOfPage: false })``
3. remove default css but remain ``position: fixed`` (the button gone while no that)
4. remove ``web-kit`` from styled-jsx
5. change button-hide style from ``display: none;`` to ``opacity: 0.01;`` add more ``pointer-events: none;``
6. pull master => forked and rebase master => branch feature
7. fixed the conflict file and jsx-a11y issues

Please review, thank you

dtinth commented on components/BacktoTopSection.js:
```diff
@@ -0,0 +1,65 @@
+            background-color: #777;
+          }
+          .button-hide {
+            opacity: 0.01;
```
Can use `opacity: 0` here.

dtinth commented on components/BacktoTopSection.js:
```diff
@@ -0,0 +1,65 @@
+            width: 30px;
+            height: 30px;
+            background: #222 url(static/arrow-up.svg) no-repeat center center;
+            background-size: 30px 20px;
```
I think the arrow looks too large; try `20px 14px`.

dtinth commented on components/BacktoTopSection.js:
```diff
@@ -0,0 +1,65 @@
+          id='top-arrow'
+          className={`top-arrow ${this.state.atTopOfPage ? 'button-hide' : ''}`}
+          onClick={this.scrollToTop}
+          onKeyDown={() => {}}
```
This is not needed since we already used `aria-hidden`.

dtinth commented on pages/index.js:
```diff
@@ -9,6 +9,7 @@ import SponsorsSection from '../components/SponsorsSection'
 import LogoBackground from '../components/LogoBackground'
 import MobileNavigation from '../components/MobileNavigation'
 import ScheduleSection from '../components/ScheduleSection'
+import BacktoTopSection from '../components/BacktoTopSection'
```
This is not a section; it is a button. Let's call it BackToTopButton

dtinth commented:
@Nutchawat I have updated your branch with the latest master from this repository. Please run `git pull` first before addressing these comments. Thank you.

Nutchawat commented:
I've pushed any change from your comments as follow.

1. update current master to feature branch ``git pull``
2. remove ``jquery`` out of ``package.json`` (sry for that)
3. fixed css about opacity ``opacity: 0;``
4. improved arrow size to ``20px 14px``
5. clean unnecessary event ``onKeyDown``
6. used ``window.scrollY`` instead of ``document.body.scrollTop`` and ``document.documentElement.scrollTop``
7. initially call ``handleScroll`` on ``componentDidMount``
8. change filename from ``BacktoTopSection`` to ``BackToTopButton``

Please try again, thank you krab.
@dtinth

Nutchawat commented:
@dtinth 
In case of pull request is approved but branch still conflicted by index.js.

Do I have to resolve right?

n3tr commented:
LGTM 👍

Nutchawat commented:
application #9SAMCP5L
